numpy/core/src/multiarray/ctors.c

```diff
@@ -3969,6 +3976,7 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
             ((PyArrayObject_fields *)ret)->data = new_data;
         }
         PyArray_DIMS(ret)[0] = i + 1;
+        PyArray_STRIDES(ret)[0] = elsize;
```

Comment:
When exactly does this matter?

Reply:
FromIter is a bit weird here (I have a PR that cleans it up a bit), but I am also confused that it should matter, since the array should be empty to begin with.

The problem is that `FromIter` sometimes resizes the array (i.e. grows it dynamically), and in that case the stride has to be sane.  (This only matters if `__length_hint__ == 0` probably)

Reply:
The PR is: gh-20993

Reply:
> This only matters if `__length_hint__ == 0`

Exactly the case that tripped this: `np.fromiter((x**2 for x in range(24)), int)`. So `__length_hint__` ==0, and the code creates a 1d array with shape `(0)` and strides (which before this PR was `elsize=8`) `(0)`. Then the code calls `next()` to get values from the iterator, expanding the result for each valid  value. Since the strides is `(0)`, the end result is a 1d array with shape `(n)` but strides `(0)`, which is wrong.

Reply:
I think this would make more sense on line 3966 when the array is reallocated, rather than doing it on every iteration.

Reply:
Rebased this on top of #20993, and moved the stride fixup to a better location (where we are sure `ret.size > 0`